File: src/0144-二叉树的前序遍历/0144.py
class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        ans = list()
        if not root:
            return ans

        stack = []
        node = root
        while stack or node:
            while node:
                ans.append(node.val)
                stack.append(node)
                node = node.left
            node = stack.pop()
            node = node.right
            
        return ans

File: src/0144-二叉树的前序遍历/0144.cpp
class Solution
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> ret;
        if (!root)
            return ret;
        deque<TreeNode *> s;
        s.push_front(root);
        while (!s.empty())
        {
            TreeNode *node = s.front();
            s.pop_front();
            ret.push_back(node->val);
            cout << " " << node->val;
            if (node->right)
            {
                s.push_front(node->right);
            }
            if (node->left)
            {
                s.push_front(node->left);
            }
        }
        return ret;
    }
};